### PR TITLE
only emit `equal` when needed: addresses #1852

### PR DESCRIPTION
**Pull request by cscheid**
**What issue does this pull request resolve?**
This addresses issue #1852 .

**What changes did you make?**
I added a check to the enum types before emitting the requirement for `equal`.

**Is there anything that requires more attention while reviewing?**
Not that I'm aware of.

### CODE REVIEW

**Comment by cscheid** [comment]
I had to add a `(schema as any[])` cast to schema.some() in order to pass `eslint`, but that seems safe since the code just above it already assumes an array with `schema.length`

**Comment by cscheid** [comment]
(Once more with feeling) An explicit check for Array.isArray, and now `npm run test` passes locally for me.

File: lib/vocabularies/validation/enum.ts [hunk 1]
```diff
@@ -20,7 +21,12 @@ const def: CodeKeywordDefinition = {
     const {gen, data, $data, schema, schemaCode, it} = cxt
     if (!$data && schema.length === 0) throw new Error("enum must have non-empty array")
     const useLoop = schema.length >= it.opts.loopEnum
-    const eql = useFunc(gen, equal)
+    const needsEql =
+      useLoop ||
+      $data ||
+      !Array.isArray(schema) ||
```
Comment (epoberezkin) [inline lib/vocabularies/validation/enum.ts]:
```suggestion
      ($data && !Array.isArray(schema)) ||
```

File: lib/vocabularies/validation/enum.ts [hunk 2]
```diff
@@ -20,7 +21,11 @@ const def: CodeKeywordDefinition = {
     const {gen, data, $data, schema, schemaCode, it} = cxt
     if (!$data && schema.length === 0) throw new Error("enum must have non-empty array")
     const useLoop = schema.length >= it.opts.loopEnum
-    const eql = useFunc(gen, equal)
+    const needsEql =
```
Comment (epoberezkin) [inline lib/vocabularies/validation/enum.ts:24]: I've made a small correction, but in general this is a brittle approach - if the condition of usage changes, these conditions would have to be updated too, and there may be edge cases that would not be caught by the tests.

A much better approach would be to simply call useFunc in the place where eql is actually used.

Comment (epoberezkin) [reply lib/vocabularies/validation/enum.ts:24]: so probably just make a memoized getEql function and replace eql with getEql() in two places.

**Comment by epoberezkin** [comment]
closing in favour of #1922
